Raise a warning for non-compliant PDF files and fix file_size validat…

qchaupds opened the pull request:
…ion for very large files

Please merge pds4-jparser first (from the other pull request) as validate depends on pds4-jparser.

Closes NASA-PDS/validate#164, NASA-PDS/validate#326, and NASA-PDS/validate#327

This pull request resolves three issues: Raise a warning if a PDF file is not PDF/A compliant and fixes file_size validation for very large files. 

Tested in DEV for #164:

% validate -R pds4.label --skip-context-validation -r report_github164_label_pdfa_invalid.json -s json  -t src/test/resources/github164/invalid/document_test_1_pdf.xml
% validate -R pds4.label --skip-context-validation -r report_github164_label_pdfa_valid.json -s json  -t src/test/resources/github164/valid/document_pdfa_valid.xml

Tested in DEV for #326 and #327 (Note that no regression was added for this issue since the file is too big to be added to the git repository):
The file is over 3GB in size and contains 38 million records.

% validate -R pds4.label --skip-content-validation --skip-context-validation  -r report_github326_label_valid_big_file_check_only.json -s json -t /data/home/pds4/qchau/test_artifacts/github326/spectra_data_collection_inventory.xml

% validate -R pds4.label --skip-context-validation -r report_github326_label_valid_big.json -s json -t /data/home/pds4/qchau/test_artifacts/github326/spectra_data_collection_inventory.xml

% validate -R pds4.label --skip-context-validation -r report_github326_label_valid_small.json -s json -t /data/home/pds4/qchau/test_artifacts/github326/spectra_data_collection_inventory_small.xml

% validate -R pds4.label --skip-context-validation -r report_github326_label_invalid.json -s json -t /data/home/pds4/qchau/test_artifacts/github326/spectra_data_collection_inventory_invalid.xml

Commit:
Raise a warning for non-compliant PDF files and fix file_size validation for very large files

1.  Add new class PDFUtil.java to validate a PDF document using new pdfbox module
2.  Add resources github164 to src/test/resources
3.  Add new external PDF validation module pdfbox and enforce same version of JAXB 2.3.0.1 in pom.xml
4.  Fix large file_size extraction of external file in FileSizesUtil.java
5.  Add function to reduce the size of informationModelVersions list in LabelUtil.java
6.  Fix building of checksum for large file in MD5Checksum.java
7.  Add type NON_PDFA_FILE to ProblemType.java
8.  Add logic to validate PDF file and modify type from int to long to handle large file size in FileReferenceValidationRule.java
9.  Add debug printing to TableDataContentValidationRule.java
10.  Add logic to reduce the size of informationModelVersions in ValidateLauncher.java
11.  Add new tests for github164 to file validate.feature
12. Correct the number of warnings for github17.   The correct number is 2 for the invalid test

Refs:

https://github.com/NASA-PDS/validate/issues/164 As a user, I want to validate PDF files are PDF/A
https://github.com/NASA-PDS/validate/issues/326 File-size check fails for large data files
https://github.com/NASA-PDS/validate/issues/327 validate fails to process large data file

jordanpadams committed:
Update POM to correct versions

validate - 2.1.0-SNAPSHOT
pds4-jparser - 2.1.0-SNAPSHOT

jordanpadams commented on src/main/java/gov/nasa/pds/tools/util/MD5Checksum.java:
```diff
@@ -54,23 +62,49 @@ public static String getMD5Checksum(URL url) throws Exception {
-    } finally {
-      input.close();
-    }
+        ByteBuffer byteBuffer = ByteBuffer.allocate(bufferSize);
```
@qchaupds similar to my comment in pds4-jparser, do we think it may make sense to group a lot of these data content checks together into one "walk" through the bits of the file? this seems like a 3rd time we are now stepping through the product to check something, which is a 3x performance degradation for very large files. it may not be possible, but worth a discussion.

jordanpadams replied:
from @qchaupds The MD5Checksum.java is a standalone utility to calculate the checksum.  It has no knowledge about the file type or record size. It should also be noted that if the `--skip-content-validation` is used, the checksum and file_size do get validated.   The checksum check is apart from the content validation.

jordanpadams replied:
will table this re-design for a later refactor

jordanpadams commented on src/main/java/gov/nasa/pds/tools/validate/rule/pds4/FileReferenceValidationRule.java:
```diff
@@ -220,16 +231,45 @@ private boolean validate(DocumentInfo xml) {
+                }
               } else if ("md5_checksum".equals(child.getLocalPart())) {
                 checksum = child.getStringValue();
+LOG.debug("FileReferenceValidationRule:validate:checksum {}",checksum);
```
@qchaupds very nitpicky here, but can we fix this whitespace if we want to keep these log messages in this method?

Commit:
Responding to sets of comments from pull requests: reformat debug printing

1.  Reformat debug printing to match indentation
2.  Add debug printing to RawTableReader.java

Refs:

https://github.com/NASA-PDS/validate/issues/164 As a user, I want to validate PDF files are PDF/A
https://github.com/NASA-PDS/validate/issues/326 File-size check fails for large data files
https://github.com/NASA-PDS/validate/issues/327 validate fails to process large data file

Commit:
Merge branch '#164_326_327' of github.com:NASA-PDS/validate into #164_326_327

Get changes from Jordan regarding the pom.xml version.
